rubber_duck.py

```diff
@@ -1,156 +1,165 @@
-                                                             usage['prompt_tokens'],
-                                                             usage['completion_tokens'])
-
-                    await self._metrics_handler.record_message(
-                        guild_id, thread_id, user_id, response_message['role'], response_message['content'])
-
-                    message_history.append(GPTMessage(role='assistant', content=response))
-
-                    await self._send_message(thread_id, response)
-
-                except Exception as ex:
-                    error_code = str(uuid.uuid4()).split('-')[0].upper()
-                    logging.exception('Error getting completion: ' + error_code)
-                    error_message = (
-                        f'😵 **Error code {error_code}** 😵'
-                        f'\nhttps://discord.com/channels/{guild_id}/{thread_id}'
-                        f'\n{ex}\n'
-                        '\n'.join(tb.format_exception(ex))
-                    )
-                    await self._report_error(error_message)
-
-                    await self._send_message(thread_id,
-                                             f'😵 **Error code {error_code}** 😵'
-                                             f'\nAn error occurred.'
-                                             f'\nPlease tell a TA or the instructor the error code.'
-                                             '\n*This conversation is closed*')
-                    return
-
-    @step
-    async def _get_completion(self, thread_id, engine, message_history) -> tuple[list, dict]:
-        # Replaces _get_response
-        async with self._typing(thread_id):
-            completion: ChatCompletion = await client.chat.completions.create(
-                model=engine,
-                messages=message_history
-            )
-            logging.debug(f"Completion: {completion}")
-            completion_dict = completion.dict()
-            choices = completion_dict['choices']
-            usage = completion_dict['usage']
-            return choices, usage
+import asyncio
+import logging
+import os
+import uuid
+import traceback as tb
+from typing import TypedDict, Protocol, ContextManager
+from openai import AsyncOpenAI
+from openai.types.chat.chat_completion import ChatCompletion
+from quest import step, queue
+
+from metrics import MetricsHandler
+
+client = AsyncOpenAI(api_key=os.environ['OPENAI_API_KEY'])
+
+AI_ENGINE = 'gpt-4'
+CONVERSATION_TIMEOUT = 60 * 3  # three minutes
+
+V_SUPPORT_STATE_COMMAND = '2023-09-26 Support State'
+V_LOG_ZIP_STATS = '2023-09-26 Zip log file, Stats'
+
+
+class Message(TypedDict):
+    guild_id: int
+    channel_name: str
+    channel_id: int
+    author_id: int
+    author_name: str
+    author_mention: str
+    message_id: int
+    content: str
+    file: list
+    
+class Attachment(TypedDict):
+    attachment_id: int
+    description: str
+    filename: str

```

Comment:
Define `Attachment` before `Message` so you can use the type for `file: list[Attachment]`

Reply:
This causes a NameError, even if I define Attachment before hand like this -> Attachment: dict    file: list[Attachment]
attempts at amending this have just led to more NameErrors even with further defintion